File: Excel/900-999/984/Challenge 984.R
library(tidyverse)
library(readxl)
library(slider)

path <- "900-999/984/984 Assigning Machine States.xlsx"
input <- read_excel(path, range = "A1:C21")
test <- read_excel(path, range = "D1:D21")

result = input %>%
  mutate(Result = recode(Result, "Pass" = 1, "Fail" = 0)) %>%
  mutate(
    rolling_sum = slide_dbl(Result, sum, .before = 2, .complete = TRUE)
  ) %>%
  mutate(
    state = accumulate2(
      Result,
      rolling_sum,
      .init = "Active",
      ~ case_when(
        ..1 == "Active" & ..3 <= 1 ~ "Halted",
        ..1 == "Halted" & ..3 == 3 ~ "Active",
        TRUE ~ ..1
      )
    )[-1]
  )

all.equal(result$state, test$`Answer Expected`)
# [1] TRUE

File: Excel/900-999/984/Challenge 984.py
import pandas as pd

path = "900-999/984/984 Assigning Machine States.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=21, skiprows=0)
test = pd.read_excel(path, usecols="D", nrows=21, skiprows=0)

states = []
for i in range(len(input)):
    prev_state = 'Active' if i == 0 else states[-1]
    if i < 2:
        states.append(prev_state)
    else:
        last_three = input.iloc[i-2:i+1]['Result']
        if prev_state == 'Active':
            states.append('Halted' if (last_three == 'Fail').sum() >= 2 else 'Active')
        else:
            states.append('Active' if (last_three == 'Pass').sum() == 3 else 'Halted')
input['State'] = states

print(input['State'].equals(test['Answer Expected']))
# True